Ayn/en 8240/more user search options

aynleslie opened the pull request:
#### What it do?

Generally speaking, this opens up many more search options for users.   As context for this PR, I highly recommend reading the new README within the /cetera-http directory first.  This explains all of the new params and gotchas.

More specifically, this:

- has the UserSearchService use the QueryParametersParser and get back `ValidatedUserQueryParameters`; this was previously a complaint Jason had that the user service didn't follow the same code path and couldn't benefit from the same error checking, merging and defaults.
- introduces the `UserSearchParamSet` akin to the other param sets we have. This is nice for consistency, but also for the benefits of encapsulation and less param passing.
-  introduces the `UserFilters` and `UserQueries` so that the growing number of user filters and queries are organized like the rest of our code.
- of course, adds new params to search by role, email, screen name and domain.
- adds new params to allow for paging.

#### How was it tested?

You'll see plenty of new tests. 
We (Ricky and I) of course checked the usual sbt things, like styleCheck and test.
We both did quite a bit of manual testing.   Ricky against staging, me against prod. 

#### Questions/Notes

1.  We felt that some things could be better named, so feel free to suggest renaming all over the place.
2. Should we also support search by `id`, i.e. 4x4?  We don't do this for /catalog and `email` is unique, so you can find a specific user that way.  But still, should we support it?
3. There are two means of providing a domain.  The X-Socrata-Host is needed to auth, but there is also a `domain` param.  I decided against using X-Socrata-Host as the domain, because we need to be able to support searches across domains, which amounts to not passing a `domain`, while still needing to pass the host header.

rlvoyer commented on cetera-http/README.md:
```diff
@@ -0,0 +1,61 @@
+The /whitepages endpoint is used to search for users.
+
+There are two important details to remember when dealing with users:
+  1. Users transcend domains.
```
Nit: this could be a little clearer. Maybe?

```
1. Users are not restricted to a particular domain.
```

aynleslie replied:
yep, will change to your line.

chrismetcalf replied:
@aynleslie What does that mean?

Its difficult to describe to end-users, because users span domains, but only within deployments.

rlvoyer commented on cetera-http/README.md:
```diff
@@ -0,0 +1,61 @@
+ 
+ * `id` is always present and is the 4x4 of the user.
+ * `screen_name` is present if it has been set by the user. Note that this
+ is also called "display_name" in some code bases; these are one in the same.
```
Nit: "code bases" => "codebases"

aynleslie replied:
really?  alright.  will change.

rlvoyer replied:
Both are acceptable, I think, but according to Google, the compound form is more common. ¯\_(ツ)_/¯

rlvoyer commented on cetera-http/README.md:
```diff
@@ -0,0 +1,61 @@
+* `q` returns any users matching the given query. Matching is based on 
+whether tokens of the user's email and screen name align with the tokens
+of the given query.  Thus for example if you use `q=funshine.bear@care.alot`,
+you will get back users who have "funshine", "bear", "care" or "alot" in
```
This example is slightly misleading now that we're filtering out frequently occurring top-level domains. You might want to include that as a parenthetical.

aynleslie replied:
ah. good thought. will do.

chrismetcalf replied:
Later in the code it looks like you've got some paging support - is this API paged?

aynleslie replied:
oh, right.  I forgot to add `offset` and `limit` and yes, we added paging!  Hard to believe, but core doesn't support that. If you add `limit`, any limit, to the /api/search/users.json query, you get no users at all, just the count of all that would have been returned.

rlvoyer commented on cetera-http/src/main/scala/com/socrata/cetera/services/UserSearchService.scala:
```diff
@@ -10,49 +10,55 @@ import org.slf4j.LoggerFactory
       }
 
-    (status, results, timings, setCookies)
+      val (results: Seq[EsUser], userSearchTime) = userClient.search(searchParams, pagingParams, domain.map(_.domainId))
```
Q: Type annotation for readability? It's not necessary, AFAICT.

aynleslie replied:
will try and remove it.

rlvoyer commented on cetera-http/src/main/scala/com/socrata/cetera/services/UserSearchService.scala:
```diff
@@ -10,49 +10,55 @@ import org.slf4j.LoggerFactory
 
-    (status, results, timings, setCookies)
+      val (results: Seq[EsUser], userSearchTime) = userClient.search(searchParams, pagingParams, domain.map(_.domainId))
+      val formattedResults = SearchResults[DomainUser](results.flatMap(u => DomainUser(domain, u)), results.size)
```
Same here re: type annotation.

aynleslie replied:
shall remove.

rlvoyer commented:
> Should we also support search by id, i.e. 4x4? We don't do this for /catalog and email is unique, so you can find a specific user that way. But still, should we support it?

Hell, yes. We should do this for catalog, too.

This looks great. Good docs, awesome tests. Minor comments. 🚢  at will.

aynleslie commented on cetera-http/README.md:
```diff
@@ -0,0 +1,61 @@
+* `X-Socrata-Host` with the domain cname
+* `Cookie` with the "_core_session_id" cookie
+
+##### Accepted Params
```
per params and param names, do you think `email`, `screen_name`, `role` and `domain` should be `emails`, `screen_names`, `roles` and `domains`?

rlvoyer replied:
That's the form we have been using so far, isn't it? So yeah, let's be consistent. Good catch.

chrismetcalf replied:
Can a user actually have multiple screen emails, screen names, roles, etc? You're describing the object, not trying to match the parameters.

aynleslie replied:
no, a user can't have multiple emails or screen names.  though I was thinking more along the lines of answering the search "show me all users with roles = viewer or publisher".  In general, when we have accepted more than 1 value for the same key, we've OR'd the values.

chrismetcalf commented on README.md:
```diff
@@ -62,6 +37,28 @@ In `/usr/local/opt/elasticsearch/config/elasticsearch.yml` set `cluster.name: ca
+### Internal endpoints
 
-Please see the [Apiary API spec](http://docs.cetera.apiary.io/#) for API details.
+See the [Internal API spec](cetera-http/README.md)
```
Having an "Internal API Spec" doesn't mean much when this whole repo is public...

aynleslie replied:
ya, Robert, Ricky and I all debated where the API spec of something internal should go.  We thought about private google docs, apiary usage and within the code base and none seemed right.   If we put it in apiary, no one, but those of us on the vpn will be able to hit it, so that's wrong.  A separate google doc would surely drift into an outdated state.  And as you say, internal routes in a public codebase is a bit pointless.   I asked myself yesterday 'What would metcalf do?'

chrismetcalf replied:
Given that the codebase is public, why is the API "private"? I'd probably instead advocate for building the right API and making it public.

aynleslie replied:
Agree with that.  Though, if we do, we should get the other internal routes on board as well.  Currently, we have 

- /whitepages (this PR)
- /personal_catalog (in the works by Ricky) that implicitly adds on filters to filter away data not owned or shared to the logged in user (and which could fairly easily be a call to /catalog, since it is a subset of the catalog)
- /asset_selector (in the works by Krah) that implicitly adds on filters to only show views that are visible at /browse or views that are owned or shared to the calling user. This too is a subset of the catalog. 

And suddenly I'm realizing, we should have talked to you sooner.

chrismetcalf replied:
:) This is why API design happens at the beginning, not the end.

Can we meet and talk about all the stuff you guys have in flight, and attempt to rationalize it in some way?

chrismetcalf commented on README.md:
```diff
@@ -62,6 +37,28 @@ In `/usr/local/opt/elasticsearch/config/elasticsearch.yml` set `cluster.name: ca
+
+* `/catalog` returns search results for query passed in parameter `q`
+* `/catalog/domains` returns a count query grouped by domain
+* `/catalog/categories` returns a count query grouped by category
```
Is this the same of the `/catalog/v1/categories` and similar endpoints?

aynleslie replied:
yes it is.  There's a note a few lines below about that.

chrismetcalf replied:
OK, then it should follow the versioned URL structure

aynleslie replied:
alright.  will do.  and that's makes it more consistent with apiary as well.

chrismetcalf replied:
Cool. Unversioned URLs are risky, since if somebody builds against that path when it maps to `v1`, and then the API changes version, their code is at risk of breakage.

hougland commented on cetera-http/src/main/scala/com/socrata/cetera/services/UserSearchService.scala:
```diff
@@ -10,49 +10,55 @@ import org.slf4j.LoggerFactory
       extendedHost: Option[String],
       requestId: Option[String])
-    : (StatusResponse, SearchResults[DomainUser], InternalTimings, Seq[String]) = {
+  : (StatusResponse, SearchResults[DomainUser], InternalTimings, Seq[String]) = {
```
nit: the alignment of this line shouldn't change

hougland commented on cetera-http/src/main/scala/com/socrata/cetera/services/UserSearchService.scala:
```diff
@@ -10,49 +10,55 @@ import org.slf4j.LoggerFactory
+        SearchResults(Seq.empty, 0),
+        InternalTimings(Timings.elapsedInMillis(now), Seq(0)),
+        setCookies
+        )
```
Refactoring out `returnUnauthorized` was the one productive thing I did today...I'll PR that tiny change maybe on monday, unless you want to incorporate it in this PR.

hougland commented on cetera-http/src/main/scala/com/socrata/cetera/types/FieldTypes.scala:
```diff
@@ -253,10 +253,19 @@ case object NameFieldType extends DocumentFieldType with Sortable {
+
+case object UserDomainId extends UserFieldType with Rawable {
+  val fieldName: String = "roles.domain_id"
+}
```
ah, you figured out we can actually specify "role_name" instead of doing the weird hack I did 🎉

hougland commented on cetera-http/src/test/scala/com/socrata/cetera/services/UserSearchServiceSpec.scala:
```diff
@@ -57,13 +57,14 @@ class UserSearchServiceSpec extends FunSuiteLike with Matchers with TestESData
     super.afterAll()
   }
 
+  println(users)
```
nit: I assume this should be removed

aynleslie commented:
Hey @rlvoyer , apologies for dragging this out, but ya know, API meeting between your last approval and now.   I've changed the following things (all in the last commit b614381, which is probably the easiest way to view the changes):

* removed the new readme and simplified the main readme to just point to apiary, which will soon contain everything the added readme had.
* pluralized the params, which prompted me to slightly refactor some things in the param parser, like
  * simplifying the `mergeParams` method to do exactly what we need and not more
  * `arrayify` strings rather than duplicate them if we want `[]` tacked onto a string.
* added the new /catalog/v1/users endpoint (for the benefit of the cetera-ruby gem and the future), but left the whitepages routes (for backwards compatibility with what's already in the frontend).

And of course, I've retested all of the routes.  Metcalf sadly hasn't looked at the apiary changes, so we lack his blessing, but still thought we could get the code review out of the way.

rlvoyer commented on cetera-http/src/main/scala/com/socrata/cetera/handlers/ParamSets.scala:
```diff
@@ -43,3 +43,15 @@ case class FormatParamSet(
+    emails: Option[Set[String]],
+    screenNames: Option[Set[String]],
+    roles: Option[Set[String]],
+    domain: Option[String],
```
Q: I feel like we discussed this, but I don't recall the conclusion. Any reason to make domain pluralable?

aynleslie replied:
we decided to leave it singular basically because of the role-domain interaction.  If  we wanted to support this, we'd either have to:

- change the return type  - to have pairs of (domain,roles) instead of role, which we aren't doing in order to stay consistent with core.  
- don't return any role information if multiple domains are past (since we wouldn't know what to return if a user had roles on 2 domains).

But now that you bring it up, we just aren't returning role info in most cases. The one and only case where you get role info is if a singular domain is passed.

So, um ... I'm not sure if we should change it.

rlvoyer replied:
Okay, I'm fine with it is.

rlvoyer commented on cetera-http/src/main/scala/com/socrata/cetera/handlers/QueryParametersParser.scala:
```diff
@@ -298,17 +326,35 @@ object QueryParametersParser { // scalastyle:ignore number.of.methods
     }
   }
+
+  def prepUserParams(queryParameters: MultiQueryParams): ValidatedUserQueryParameters = {
```
👍  so clean

rlvoyer commented on cetera-http/src/main/scala/com/socrata/cetera/handlers/QueryParametersParser.scala:
```diff
@@ -298,17 +326,35 @@ object QueryParametersParser { // scalastyle:ignore number.of.methods
+       prepareUserDomain(queryParameters),
+       prepareUserQuery(queryParameters)
+     )
+     val pagingParams = PagingParamSet(
```
Should we give `sortOrder` a default value of None? It would make things slightly cleaner maybe?

aynleslie replied:
yes, I think that makes much more sense.  Will do.

rlvoyer commented on cetera-http/src/main/scala/com/socrata/cetera/handlers/QueryParametersParser.scala:
```diff
@@ -407,22 +460,22 @@ object Params {
-    stringKeys.contains(key) ||
-      arrayKeys.contains(key) ||
-      key.endsWith("]") && mapKeys.contains(key.split('[').head)
+    catalogKeys.contains(key) ||
```
Nit: odd indentation

aynleslie replied:
actually, since you bring it up, what do we prefer in this case?  I didn't change the indentation here, but it did catch my eye too.  I tend to prefer the long booleans lined up, but intellij defaults to tabbing over subsequent lines and I can't find any guidelines on line.

rlvoyer replied:
Actually, digging in a bit more, while this may look odd to my eyes, this conforms w/ the official Scala style guide and there is no mention of how to handle line continuation in the Spark style guide, that I can see. Sorry for the false alarm.

rlvoyer commented on cetera-http/src/main/scala/com/socrata/cetera/search/Queries.scala:
```diff
@@ -215,3 +216,25 @@ object DocumentQueries {
 }
+
+object UserQueries {
+  def emailNameMatchQuery(query: Option[String]): QueryBuilder = {
```
Nit/FYI: you don't need outer function braces because the function body is a single expression

aynleslie replied:
oh, yep.  I tend not to notice that when it's a multi-line single expression.  thanks for pointing out - will edit.

rlvoyer commented on cetera-http/src/main/scala/com/socrata/cetera/types/FieldTypes.scala:
```diff
@@ -253,10 +253,23 @@ case object NameFieldType extends DocumentFieldType with Sortable {
 // U'sarians
 
-case object ScreenName extends UserFieldType with Rawable {
+case object UserId extends UserFieldType with Rawable {
```
I don't think user ID is rawable because there's no "raw" subfield. I think this should extend NativelyRawable, right?

aynleslie replied:
oh, true, true.  And yet this worked fine.  You'd kind of expect things to break if the traits are just wrong.  sigh ...  regardless, will change and retest.

rlvoyer replied:
I think nothing failed because you're not referencing the `rawFieldName` field in the new queries/filters. So you unnecessarily mixed in this trait presumably because you were following the patterns you observed in the code. It's not well documented what those traits provided us. We should fix that in a separate commit.

rlvoyer commented on cetera-http/src/main/scala/com/socrata/cetera/types/FieldTypes.scala:
```diff
@@ -253,10 +253,23 @@ case object NameFieldType extends DocumentFieldType with Sortable {
+  val fieldName: String = "roles.role_name"
+}
+
+case object UserDomainId extends UserFieldType with Rawable {
```
This also isn't rawable.

aynleslie replied:
and neither is the role.  only the email and screen_name are.  Just sloppy on my part.  apologies. will fix.

though again, would help if our traits/types broke something when wrong.  Probably insufficient code coverage.  and me being sloppy.

rlvoyer commented:
Aside a from a few comments, this looks really good to me. Address the Rawable field types issue and I say :shipit:

aynleslie commented:
manually merged. thank you for the double reviews robo.
